### DISCUSSION TITLE
dynamic component provider - per component consistent providing of attributes

### DISCUSSION THREAD

**Discussion by RonnyPfannschmidt** [reagento/dishka#258] [Ideas]
in a Application framework im working on i want to migrate towards using dishka to provide data of plugins

however instead of having something like

```python
...

class PluginProvider(Provider):
   ...


plugin_providers = [PluginProvider(component=plugin_name) for plugin_name in get_availiable_plugins()]
```

i want to use something more like 

```python

...

class PluginProvider(DynamicComponenProvider):
  ...


plugin_provider = PluginProvider()

````

if necessary i can extract a more complete example to demonstrate the point

**Comment by Tishka17** [comment]
You can inherit from BaseProvider and make your own login on filling lists inside it

**Comment by RonnyPfannschmidt** [comment]
I saw how providerwrapper is implemented and off hand its not clear to me how to propperly map the component of Depend objects 

As looks to me like component is taken from the Provider unless i manually create All components in the dependencies 

Using providerwrapper may be the most prachical solutikn for now

**Comment by Tishka17** [comment]
`Component` from factory is used when it is not set explicitly in dependency. But, if you are creating new BaseProvider from existing `Provider` you can use `with_component` method on each factory to fill it. It is a bit more complicated on decorators and aliases, but not very much. I guess we can extend their API adding same method.

**Comment by RonnyPfannschmidt** [comment]
for my immediate needs, im under the impression, using ProviderWrapper is a more sensible approach as starting point

as there is a need to replicate all providers/dependency keys its much better to make instances per component

**Comment by RonnyPfannschmidt** [comment]
@Tishka17 i believe i just found the problem - there is seemingly no access to the correct self.component value in provide methods
